### DISCUSSION TITLE
V3 progress

### DISCUSSION THREAD

**Discussion by WhyAsh5114** [WhyAsh5114/MyFit#45] [Announcements]
Many people want to know about the progress of V3 and when I plan to release it. I won't be able to give a fixed release date as that forces me to deliver an incomplete product like I did with V2, you can read more about that [here.](https://my-fit-whyash5114.vercel.app/v3_blog#about_v3)

I'll create a progress list here to avoid keeping everyone in complete darkness and update it regularly, I've removed most of the completed items from this to condense it like ExerciseSplit and Mesocycle stuff:

# Working on right now
- Workouts
    - [x]  add and edit exercise
    - [x]  modify mesocycle exercise split days according to workouts
    - [x]  different set UIs for setTypes and modularize ExerciseCard
        - [x] straight
        - [x] v2
        - [x] drop
        - [x] myorep
        - [x] myorep match
        - [x] down/top
    - [x]  create workout endpoint
    - [x]  scaling bodyweights (1/3 for push ups, 1 for pull ups, etc.)
    - [x]  Temporary exercise skipping
    - [x]  ESLint CI/CD job
    - [x]  workouts list view page (basic)
    - [x]  workout view page
    - [x]  edit and delete workout
    - [x] workout skipping
    - [x]  basic tests
    - [x]  overview page
    - [x]  compare action button implementation
    - [x]  Prefill exercise rep, load, RIR values when adding new ones
    - [x]  notifications! [card example, scroll to the bottom](https://www.shadcn-svelte.com/docs/components/card)
        - [ ] persistent workout time
        - [x] card in settings and push notifications permission stuff
        - [ ] options to enable/disable particular types of notifications
    - [x]  Progressive overload function...
    - [x] 1 RM calculation when changing rep/load values
    - [ ]  comprehensive tests
    - [x]  workouts list view page filters

# Soon
- [ ]  PWA not working properly: offline support and caching should be tested thoroughly
- [x]  posthog for analytics
- [ ]  Mesocycle end options: extend, deload, finish
- [ ]  Automatic deload creation
- [ ]  Thorough tests for exercise splits and mesocycles, (start immediately, cut, copy, paste, progression overrides, swap split day exercises, warnings, errors, etc.)
- [x]  Migrate to cuid2, cuid package is not secure https://github.com/prisma/prisma/issues/17102
- [ ]  manual update & app version section in settings
- [ ]  Offline indicator icon
- [x]  Clear local storage runes whenever PWA updates
- [ ]  Post activation potentiation [Menno's post](https://www.instagram.com/p/C8UcREYulnO/?igsh=MWwxb2drc2x4Znow)
- [ ]  DOCS! each info popover should link to a docs section?
- [ ]  Consider wrapping Prisma queries in try-catch, errors do happen like can’t reach database and transaction failures [Prisma Docs](https://www.prisma.io/docs/orm/reference/error-reference)
- [ ]  Helper popovers in exercise-splits
- [ ]  A zone mode for mesocycle cyclic set changes that has levels of specialization each with its own setChange property for volume distribution, can be represented in DB with optional zone property
- [ ]  Lengthened partials should be set by using -negative RIR, -2 RIR means 2 lengthened partials?
- [ ]  Shallow routing for drawers
- [ ]  Also proper routing for `Previous` and `Next` buttons during big form creations. Using previous and then hitting back, does weird stuff, make it more intuitive
- [x]  Make sure logging out and logging in again doesn’t leak previous user info
- [ ]  motivational quotes pre, intra, and post workouts?
- [ ]  Web metrics [Fireship video](https://www.youtube.com/watch?v=0fONene3OIA)
- [ ]  Migration function from v2 (maybe v1)
- [ ]  Index creation (cockroachDB gives recommendations in the databases section), also checkout insights section
- [x]  Home page (meso info, today’s workout, stats)
- [ ]  Settings (preferences like weight increments?)
- [ ]  Profile (workout stats and stuff)
- [ ]  Checkout GitHub issues of repo for feature requests https://github.com/WhyAsh5114/MyFit/issues
- [ ]  Offline testing for saving workouts offline
- [x]  Allow workout creation without a mesocycle (with option to manually select a reference workout. With a mesocycle, reference workout will be set automatically)
- [ ]  Weight units (lb and kg)
- [ ]  Exercise swapping, when you change gyms and stuff, you can swap out exercises, a swappedOut state for the swappedOut exercise maybe?
- [x]  mesocycle workouts tab, progression, list, and a lot of stuff
- [x]  speed up /workouts/new/start and /exercises page, it takes quite the time to load
- [ ]  backgroundSyncPlugin queue UI and sync implementation (show pending requests in UI)
- [ ]  Migrate to shadcn charts?
- [ ]  Replace `onMount` functions with `$effect` to refresh page when page data changes
- Data entry
    - [ ]  exercises
    - [ ]  splits
    - [ ]  helper popovers

# Pre-production
- [x]  A proper CI/CD pipeline with database migrations, simulate workflows locally, see fireship video https://youtu.be/yfBtjLxn_6k
- [x]  GitHub sponsorships (in repo settings)
- [x]  About section with donation options in the app itself
- [x]  Privacy policy, user consent, update OAuth providers and more if wanted?
- [ ]  https://web.dev/articles/building/an-adaptive-favicon#markup
- [ ]  Publish on GitHub apps and Play store
- [x]  Complete svelte 5 migration, enable runes mode in compiler option
- [x]  Move other vercel projects to different vercel account
- [x]  New README, contributing guidelines, code of conduct
- [x]  database backups, a migration model

# Maybe
- [ ]  Re deploy v1
- [ ] use bun instead of npm
- [ ]  Help migrate svelte-sonner and cmdk-sv to svelte 5
- [ ]  i18n
- [ ]  Monetization? Ads model (show ads and go ad free with premium), freemium model (100 workouts free), donations only?

# Post production refactors
- [ ] use [builder design pattern](https://refactoring.guru/design-patterns/builder) to build the create query of complex nested objects on the frontend itself. To ensure the create query is safe, use `EntityUncheckedCreateWithoutUserInput` to avoid unauthorized user creation
- [ ] separate create, edit, and view pages, avoid using `manage` for both create and edit, too many if conditions
- [ ] use svelte query to manage loading and error states
- [ ] wrap every possible error in trycatch
- [ ] change bodyweight fraction from null | number to just number, 0 by default instead of null

**Comment by yzee-tech** [comment]
Are there things other people could help you with? For monetization, I think you should do a paypal link!

**Reply by WhyAsh5114** [reply]
It depends on your experience with svelte and the rest of the tech stack. You can help me with either super basic stuff like the data entry section, or even implement your own feature in a PR. Just a suggestion before starting to create a new feature, try to discuss it here so you can make sure that it'll be used and implemented appropriately.

**Reply by yzee-tech** [reply]
Not too familiar with svelte, is there an intro/tutorial that you recommend?

**Reply by WhyAsh5114** [reply]
You can start with their official tutorial: https://learn.svelte.dev/tutorial/welcome-to-svelte
Although keep in mind that I'm currently using Svelte 5 which is in release candidate and should be released soon, there are a few differenced in both but they are both inter-compatible, Svelte 4 code will (mostly) work with Svelte 5.
